### ISSUE TITLE
Question: How to use parameterized routes and wildcard routes

### ISSUE THREAD

**Issue by pmooneyarchitech** [swaechter/angularj-universal#21]
There's mention in the documentation of using wildcard routes but I'm not sure how this is done. Ideally I just want a simple one method controller to blanket handle all of my routes, but barring that even individual parameterized route examples would be helpful.

For example I have the route `/person/:personId` in my Angular app. In the controller in my Spring app I'm trying 

    @GetMapping({"/person/{personId}"})
    public ModelAndView showPersonDetail(HttpServletRequest request, @PathVariable("person") String personId) {
        return new ModelAndView("/person/:personId");
    }

and

    @GetMapping({"/person/{personId}"})
    public ModelAndView showPersonDetail(HttpServletRequest request, @PathVariable("person") String personId) {
        return new ModelAndView("/person/" + personId);
    }

With
`angularjuniversal.routes=/,/person/:personId`

In my application.properties file. Neither of them seem to be able to render the page on the server side with the data retrieved.

Edit: Updated for markdown formatting

**Comment by pmooneyarchitech** [comment]
It seems to me, this doesn't work in the current 0.0.2 release. The latest in master of the class [AngularJUniversalView.java](https://github.com/swaechter/angularj-universal/blob/master/angularj-universal-spring-boot-starter/src/main/java/ch/swaechter/angularjuniversal/springboot/starter/AngularJUniversalView.java) has the fix that works for me.

My solution is to provide that newer AngularJUniversalView, and AngularJUniversalViewResolver implementations in my code and override the ViewResolver bean in spring with my version of AngularJUniversalViewResolver.

After that is done, I only need to resolve one route in my application.propeties: `angularjuniversal.routes=/` and then all of my ModelAndViews go to that route. Eg,

    @GetMapping({"/"})
    public ModelAndView showHome() {
        return new ModelAndView("/");
    }

    @GetMapping({"/person/{personId}"})
    public ModelAndView showPersonDetail() {
        return new ModelAndView("/");
    }

**Comment by swaechter** [comment]
Hey Paul

Indeed, this won't work at the moment. The problem is, that the values from the route property `angularjuniversal.routes` are split to a string list and compared (with equals) to the called route. This will of course not work for paramterized routes and we have to use wildcards. For more information see https://github.com/swaechter/angularj-universal/blob/master/angularj-universal-spring-boot-starter/src/main/java/ch/swaechter/angularjuniversal/springboot/starter/AngularJUniversalViewResolver.java#L57

I don't like how we have to duplicate Spring endpoints, so I invested some time yesterday to get completely rid of that and do the routing on the property values (The endpoints are dynamically added based on the content of `angularjuniversal.routes`) - including the possibility to use wildcards like `/person/**`.

I will polish the code and try to upload it this evening/tomorrow morning and give you an update.

Best,
Simon

**Comment by swaechter** [comment]
Hey there

I just committed my changes. What did change:

1.) I removed the login system in the example and replaced it with a `/keywords` and `/keywords/:id` page. At the moment the specific keyword page returns a dummy keyword, but this should reflect your use case.

2.) As mentioned in the last comment, the `/keywords/:id` issue was related to the simple route detection. The new route detection now uses `Ant Path Matcher` that allows some regex-like behaviour like `/keywords/**`. For more information see https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/util/AntPathMatcher.html

The new code can be found here: https://github.com/swaechter/angularj-universal/blob/master/angularj-universal-spring-boot-starter/src/main/java/ch/swaechter/angularjuniversal/springboot/starter/AngularJUniversalViewResolver.java#L59

3.) The Spring Boot starter integration finally registers request mappings dynamically. You can define all "valid" Angular SPA routes via `angularjuniversal.routes` in your `application.properties` (Or wherever) and the starter will register these mappings (Like we did with `GetMapping` and `ModelAndView`. Example:

```
angularjuniversal.routes=/,/home,/keywords,/keywords/**,/about
```

Valid URIs:

```
/
/home
/keywords
/keywords/1 --> A wildcard for this is defined, but the original client URI is passed to the renderer
/about
```

You can see the loaded request mappins in the Spring Boot log:

```
2018-03-03 16:32:25.224  INFO 15586 --- [  restartedMain] o.s.w.s.handler.SimpleUrlHandlerMapping  : Root mapping to handler of type [class org.springframework.web.servlet.mvc.ParameterizableViewController]
2018-03-03 16:32:25.224  INFO 15586 --- [  restartedMain] o.s.w.s.handler.SimpleUrlHandlerMapping  : Mapped URL path [/home] onto handler of type [class org.springframework.web.servlet.mvc.ParameterizableViewController]
2018-03-03 16:32:25.224  INFO 15586 --- [  restartedMain] o.s.w.s.handler.SimpleUrlHandlerMapping  : Mapped URL path [/keywords] onto handler of type [class org.springframework.web.servlet.mvc.ParameterizableViewController]
2018-03-03 16:32:25.224  INFO 15586 --- [  restartedMain] o.s.w.s.handler.SimpleUrlHandlerMapping  : Mapped URL path [/keywords/**] onto handler of type [class org.springframework.web.servlet.mvc.ParameterizableViewController]
2018-03-03 16:32:25.224  INFO 15586 --- [  restartedMain] o.s.w.s.handler.SimpleUrlHandlerMapping  : Mapped URL path [/about] onto handler of type [class org.springframework.web.servlet.mvc.ParameterizableViewController]
2018-03-03 16:32:25.234  INFO 15586 --- [  restartedMain] o.s.w.s.handler.SimpleUrlHandlerMapping  : Mapped URL path [/webjars/**] onto handler of type [class org.springframework.web.servlet.resource.ResourceHttpRequestHandler]
2018-03-03 16:32:25.234  INFO 15586 --- [  restartedMain] o.s.w.s.handler.SimpleUrlHandlerMapping  : Mapped URL path [/**] onto handler of type [class org.springframework.web.servlet.resource.ResourceHttpRequestHandler]
2018-03-03 16:32:25.262  INFO 15586 --- [  restartedMain] o.s.w.s.handler.SimpleUrlHandlerMapping  : Mapped URL path [/**/favicon.ico] onto handler of type [class org.springframework.web.servlet.resource.ResourceHttpRequestHandler]
2018-03-03 16:32:25.288  INFO 15586 --- [  restartedMain] o.s.b.a.w.s.WelcomePageHandlerMapping    : Adding welcome page: class path resource [public/index.html]
```

**Important note**: At the moment you have to [still] define the `/` endpoint manually. All others endpoint can be deleted, but this one has to stay due some Spring Boot internal resource handling - otherwise you will see an unrendered index.html. So keep this code:

```java
@Controller
@RequestMapping("/")
public class ContentController {

    @GetMapping("/")
    public ModelAndView showIndex() {
        return new ModelAndView("/");
    }
}
```

I hope I will be able to fix this annoyance soon.

3.) The `RenderConfiguration` now uses a fluent-API design. For more information see the example in the README.md

```java
RenderConfiguration configuration = new RenderConfiguration.RenderConfigurationBuilder(templatecontent, serverbundlefile).engines(4).liveReload(false).build();
```

Can you clone the project and test the latest master commit? If it is working for you and fulfills your needs, I'll make a 0.0.3 release.

Best,
Simon

**Comment by swaechter** [comment]
Hey @pmooneyarchitech any update on this? :)

**Comment by pmooneyarchitech** [comment]
Hey @swaechter Sorry I've been away for a bit. Either myself, or someone on my team will try it out in the next day or two and I'll get back to you.

**Comment by pmooneyarchitech** [comment]
I had checked out and built the the latest code (giving it a new version number) from commit `e0bf66788c9cd1e9393fd3992cee431e08bb6bb1`. Unfortunately it gave me an error on startup. I'll try to reproduce and post back with the stacktrace if it's not easily reproducible.

**Comment by dongmei-cao** [comment]
@swaechter I just tried the latest code. It worked perfectly for me.

**Comment by swaechter** [comment]
@dongmei-cao Thank you very much for the feedback. Based on it I just released version `0.0.3`: https://github.com/swaechter/angularj-universal/releases/tag/0.0.3

@pmooneyarchitech As @dongmei-cao mentioned, I switched to Spring Boot 2 after the 0.0.2 release. I never tested how the starter (based on Spring Boot 2) behaves with Spring Boot 1 - but I guess it won't work at all.

Maybe you can provide a minimal working example so I can check it out?

**Comment by swaechter** [comment]
By the way: The Spring Boot starter can now also register the `/` route, so you don't have to provide a `ModelAndView` for this route.

This problem was fixed in Spring Boot 2.0.1: https://github.com/spring-projects/spring-boot/issues/12335
